### PR TITLE
feat: files upload integration with xblock handlers

### PR DESCRIPTION
**Pull request by johnvente**
integration backend for upload images to the XBlock

**DEMO**

STUDIO
images-studio

**LMS**
updatadde-lms

## Configure the XBlock

<table>
  <thead>
    <tr>
      <th>Step</th>
      <th>Description</th>
      <th>Screenshot URL</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>Go to <strong>"Settings"</strong> -> <strong>Advanced Settings</strong></td>
      <td>advanced-setting</td>
    </tr>
    <tr>
      <td>2</td>
      <td>Add to <strong>Advanced Module List</strong> -> "imagesgallery"</td>
      <td>setting</td>
    </tr>
    <tr>
      <td>3</td>
      <td>Add a new unit with the XBlock in <strong>Advanced</strong></td>
      <td>
        components<br>
        componentadada
      </td>
    </tr>
  </tbody>
</table>

### HOW TO TEST IT
**[tutor]**
clone this branch in your  requirements

**1.** 
```
 cd /env/build/openedx/requirements/xblock-imagesgallery && git clone -b jv/files-upload-fix-front-back git@github.com:eduNEXT/xblock-imagesgallery.git
 ```

- -----  lms -------
1. tutor dev exec lms bash
2. pip install -e /openedx/requirements/xblock-imagesgallery
3. exit 

- -----  cms -------
1. tutor dev exec cms bash
2. pip install -e /openedx/requirements/xblock-imagesgallery
3. exit 

**2.** tutor config save

**3.** tutor dev restart

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

### CODE REVIEW

**Comment by BryanttV** [comment]
When I upload an image to the component and then delete it from Studio's file manager (`Content → Files & Uploads`), the reference to the image still appears in the component.
image

**Comment by johnvente** [comment]
> When I upload an image to the component and then delete it from Studio's file manager (`Content → Files & Uploads`), the reference to the image still appears in the component. image

Yep, that happens currently because we are saving the images in [localStorage](https://developer.mozilla.org/es/docs/Web/API/Window/localStorage) and not with XBlock state, so that can't be fixed until we share the XBlock's state

File: react-app/App.jsx [hunk 1]
```diff
@@ -3,28 +3,31 @@ import DropZoneFile from '@components/DropZoneFile';
...
 
 const App = () => {
-  const [isGalleryOpened, setIsGalleryOpened] = useState(false);
   const [filesToUploadList, setFilesToUploadList] = useState([]);
+  const { xblockId, isStudioView } = globalObject;
 
   useEffect(() => {
-    localStorage.clear();
+    const filesUploaded = getItemLocalStorage(xblockId) || [];
+    setFilesToUploadList(filesUploaded);
```
Comment (mariajgrimaldi) [inline react-app/App.jsx:18]: This is good enough for now, but remember we should remove this once we save the list as part of the xblock state.

File: react-app/components/DropZoneFile/index.jsx [hunk 2]
```diff
@@ -19,43 +20,68 @@ const fileTypesAllowed = {
...
+         const data = {
+            current_page: 0,
+            page_size: pageSize
+          };
+
+        //const filesResponse = await apiConfig.post(fileGetterHandler, data);
+        // const {data: uploadedFiles } = filesResponse.files; */
+
+        // Handle the response here
+        // console.log(filesResponse.data);
```
Comment (mariajgrimaldi) [inline react-app/components/DropZoneFile/index.jsx]: not sure what this means

File: react-app/components/DropZoneFile/index.jsx [hunk 3]
```diff
@@ -19,43 +20,68 @@ const fileTypesAllowed = {
 const DropZoneFile = () => {
   const { setFilesToUploadList } = useContext(GalleryContext);
 
+  async function uploadAndFetchFiles(formData, _ = 10) {
```
Comment (mariajgrimaldi) [inline react-app/components/DropZoneFile/index.jsx]: what is the _ = 10 for?

File: react-app/components/DropZoneFile/index.jsx [hunk 4]
```diff
@@ -19,43 +20,68 @@ const fileTypesAllowed = {
...
+      const fileUploadHandler = globalObject.runtime.handlerUrl(globalElement, 'file_upload');
+
+      // Upload the file
+      const uploadResponse = await apiConfig.post(fileUploadHandler, formData, {
+        headers: {
+          'Content-Type': 'multipart/form-data'
+        }
+      });
+
+      if (uploadResponse.status === 200) {
```
Comment (mariajgrimaldi) [inline react-app/components/DropZoneFile/index.jsx]: can we use constants for the status codes? what if it fails? can we show an error message?

File: react-app/components/Gallery/index.jsx [hunk 5]
```diff
@@ -1,42 +1,89 @@
...
+
+  const handleOnThumbnailClick = (_, index) => {
+    setCurrentIndex(index);
+  };
+
+  const handleOnSlide = (currentIndex) => {
+    setCurrentIndex(currentIndex);
+  };
+
+  const fetchUploadedFiles = async (page = 0) => {
```
Comment (mariajgrimaldi) [inline react-app/components/Gallery/index.jsx:43]: what about using this type of in-line documentation? https://github.com/openedx/frontend-app-course-authoring/blob/6f238bdbe0bfe19da6a8fef5b9f05f0e9b30bb46/src/pages-and-resources/data/api.js#L26-L31

File: react-app/components/Gallery/index.jsx [hunk 6]
```diff
@@ -1,42 +1,89 @@
...
+      const formatImages = files.map(({ external_url }) => ({
+        original: external_url,
+        thumbnail: external_url
+      }));
+
+      setSizeItems(sizeItems);
+      setImagesList((prevImages) => [...prevImages, ...formatImages]);
+      setFetchedPages({ ...fetchedPages, [page]: true });
+    } catch (error) {
+      console.log('files error', error);
```
Comment (mariajgrimaldi) [inline react-app/components/Gallery/index.jsx]: can we do a better error messaging instead of console.log?

**Comment by mariajgrimaldi** [comment]
Hi there, folks @01001110J @jignaciopm! I mentioned you here if you have time to review the code. You don't need to test it out, we got it! Thanks

File: react-app/components/DropZoneFile/index.jsx [hunk 7]
```diff
@@ -19,43 +20,68 @@ const fileTypesAllowed = {
...
+            page_size: pageSize
+          };
+
+        //const filesResponse = await apiConfig.post(fileGetterHandler, data);
+        // const {data: uploadedFiles } = filesResponse.files; */
+
+        // Handle the response here
+        // console.log(filesResponse.data);
+      } else {
+        console.error('File upload failed');
```
Comment (01001110J) [inline react-app/components/DropZoneFile/index.jsx]: As user, exist any kind of visual feedback about this error?

File: react-app/components/Gallery/index.jsx (hunk 5 above)
Comment (01001110J) [reply react-app/components/Gallery/index.jsx:43]: Just in case:

https://jsdoc.app/

File: react-app/components/Gallery/index.jsx [hunk 8]
```diff
@@ -1,42 +1,89 @@
...
+  const isFirstPageFetched = Boolean(fetchedPages[0]);
+  const emptyImagesMessage = gettext('There are not images available');
+
+  if (isFirstPageFetched && !sizeItems) {
+    return <p>{emptyImagesMessage}</p>;
+  }
 
   return (
-    <div className={`gallery-container ${isGalleryOpened ? 'show' : ''}`}>
+    <div className={`gallery-container`}>
```
Comment (01001110J) [inline react-app/components/Gallery/index.jsx]: Why this format? It's kinda weird.

```suggestion
    <div className="gallery-container">
```

File: react-app/components/ListImages/components/ImageItem/index.jsx [hunk 9]
```diff
@@ -22,22 +24,26 @@ function ImageItem(props) {
...
+      const fileDeleteHandler = globalObject.runtime.handlerUrl(globalElement, 'remove_files');
+      const data = { asset_key: assetKey };
+      await apiConfig.post(fileDeleteHandler, data);
+      const filesToUploadListStorage = getItemLocalStorage(xblockId) || [];
+      const filesToUploadListUpdated = filesToUploadListStorage.filter(({ id }) => id !== idImageItem);
+      setItemLocalStorage(xblockId, filesToUploadListUpdated);
+      setFilesToUploadList(filesToUploadListUpdated);
+
+    } catch (error) {
+      console.log('it failed to delete file', error);
```
Comment (01001110J) [inline react-app/components/ListImages/components/ImageItem/index.jsx]: Same comment about the user feedback :thinking:

File: react-app/components/Gallery/styles.css [hunk 10]
```diff
@@ -7,6 +7,33 @@
 .gallery-content {
   display: flex;
   align-items: center;
+  justify-content: center;
+}
+
```
Comment (01001110J) [inline react-app/components/Gallery/styles.css:12]: It's possible to add ( in the future ) a code [formatter](https://prettier.io/)?

I've seen many css files with different types of formatting, which is not consistent :thinking:

**Comment by mariajgrimaldi** [comment]
@johnvente: you can change the base branch to main, and rebase with it

File: react-app/components/Gallery/styles.css (hunk 10 above)
Comment (johnvente) [reply react-app/components/Gallery/styles.css:12]: Actually I'm using prettier, I will add [eslint](https://eslint.org/) for some rules (not in this pr)

File: react-app/components/DropZoneFile/index.jsx (hunk 7 above)
Comment (johnvente) [reply react-app/components/DropZoneFile/index.jsx]: Not yet, on it

File: react-app/components/Gallery/index.jsx (hunk 5 above)
Comment (johnvente) [reply react-app/components/Gallery/index.jsx:43]: I do not like jsdoc, it says which types need to be passed but someone can send another type but I will use it in this case

**Comment by johnvente** [comment]
Hey @mariajgrimaldi @01001110J I've done the changes

File: react-app/components/DropZoneFile/index.jsx [hunk 11]
```diff
@@ -17,56 +20,84 @@ const fileTypesAllowed = {
...
+        setFilesToUploadList(filesUnloaded);
+        setItemLocalStorage(xblockId, filesUnloaded);
+        setDropZoneErrorMessage(null);
+      } else {
+        filesToUploadFailedMessage = gettext('File upload failed');
+        setDropZoneErrorMessage(filesToUploadFailedMessage);
+      }
+    } catch (error) {
+      console.error('Error:', error);
+      filesToUploadFailedMessage = gettext('It occurred an unexpected error');
```
Comment (mariajgrimaldi) [inline react-app/components/DropZoneFile/index.jsx]:
```suggestion
      filesToUploadFailedMessage = gettext('An unexpected error occurred);
```

**Review by mariajgrimaldi — APPROVED** [review]
This is working as expected! Thanks

**Pull request merged**